tests/tezos-modules.nix

```diff
@@ -0,0 +1,67 @@
+# SPDX-FileCopyrightText: 2021 TQ Tezos <https://tqtezos.com/>
+#
+# SPDX-License-Identifier: LicenseRef-MIT-TQ
+let
+  nixpkgs = (import ../nix/nix/sources.nix).nixpkgs;
+  pkgs = import ../nix/build/pkgs.nix {};
+in import "${nixpkgs}/nixos/tests/make-test-python.nix" ({ ... }:
+{
+  machine = { ... }: {
+    virtualisation.memorySize = 1024;
+    virtualisation.diskSize = 1024;
+
+    nixpkgs.pkgs = pkgs;
+    imports = [ ../nix/modules/tezos-node.nix
+                ../nix/modules/tezos-signer.nix
+                ../nix/modules/tezos-accuser.nix
+                ../nix/modules/tezos-baker.nix
+                ../nix/modules/tezos-endorser.nix
+              ];
+
+    services.tezos-node.instances.edonet.enable = true;
+
+    services.tezos-signer.instances.edonet = {
+      enable = true;
+      networkProtocol = "http";
+    };
+
+    services.tezos-accuser.instances.edonet = {
+      enable = true;
+      baseProtocol = "008-PtEdo2Zk";
+    };
+
+    services.tezos-baker.instances.edonet = {
+      enable = true;
+      baseProtocol = "008-PtEdo2Zk";
+    };
+
+    services.tezos-endorser.instances.edonet = {
+      enable = true;
+      baseProtocol = "008-PtEdo2Zk";
+    };
+
+  };
+
+  testScript = ''
+    start_all()
+
+    services = [
+        "tezos-node",
+        "tezos-signer",
+        "tezos-baker",
+        "tezos-accuser",
+        "tezos-endorser",
+    ]
+
+    for s in services:
+        machine.wait_for_unit(f"tezos-edonet-{s}.service")
+
+    machine.wait_for_open_port(8732)
+    machine.wait_until_succeeds(
+        "curl --silent http://localhost:8732/chains/main/blocks/head/header | grep level"
+    )
+
+    for s in services:
+        output = machine.succeed(f"systemctl status tezos-edonet-{s}.service | grep Active")
```

Comment:
I think this can just run `systemctl status` service, the exit code should be non-zero if the service fails

Reply:
Actually, just `machine.succeed(f"systemctl status tezos-edonet-{s}.service")` inevitably fails with an outcome dependent on the order of names in the `services` list: usually with `Exception: command `systemctl status tezos-edonet-tezos-baker.service` failed (exit code 3)`, but sometimes it's this or any other service with exit code 2. The thing is, the trace also shows a perfectly normal `systemctl status` output for the offending service, and then fails for no discernible (to me) reason. E.g.:

```
machine: must succeed: systemctl status tezos-edonet-tezos-baker.service
machine: output: ● tezos-edonet-tezos-baker.service - Tezos baker
     Loaded: loaded (/nix/store/3lkpyl7bldfqrs6q0342xd764h7g538n-unit-tezos-edonet-tezos-baker.service/tezos-edonet-tezos-baker.service; enabled; vendor preset: enabled)
     Active: activating (auto-restart) since Mon 2021-03-29 05:04:32 UTC; 7s ago
    Process: 931 ExecStart=/nix/store/d5d9hrzv8ampx4c4rw6zpfvkyb2rjf35-unit-script-tezos-edonet-tezos-baker-start/bin/tezos-edonet-tezos-baker-start (code=exited, status=0/SUCCESS)
   Main PID: 931 (code=exited, status=0/SUCCESS)
         IP: 0B in, 0B out
        CPU: 373ms

Mar 29 05:04:32 machine systemd[1]: tezos-edonet-tezos-baker.service: Succeeded.
Hint: Some lines were ellipsized, use -l to show in full.

Test "service status sanity check" failed with error: "command `systemctl status tezos-edonet-tezos-baker.service` failed (exit code 3)"
```

Any systemctl status run on its own (e.g. just `machine.succeed("systemctl status tezos-edonet-tezos-baker.service")`) never fails, as far as I can tell. Any idea what the problem could be?

Reply:
I don’t know why it fails, but nevertheless I want to share a piece of knowledge that I happen to possess: there is `systemctl is-active <unit>` 🤯.

Reply:
I think this actually makes sense. In your output the unit is `activating`, i.e. it is not actually running yet, thus `systemctl status` exits with a non-zero exit code.

If you want to make sure that it has actually successfully started, you’ll have to wait for it.

Also see this: https://unix.stackexchange.com/questions/385454/systemd-check-that-a-service-is-not-running. There is a difference between `systemctl status` and `systemctl is-active`, and they also mention `systemctl show -p ActiveState` – make sure you pick the right one.

Reply:
Also, as a matter of fact, `Active: activating (auto-restart) ` looks pretty sus 🤨. Is the service crashing and gets auto-restarted? :)

Reply:
_Oh, ok, I actually read the discussion above. Looks like the tests caught an actual bug? :)_

Reply:
Less of a bug than something I didn't manage to get an answer to in time from @rvem. But all good now, I think, found the culprit :tada: